pkg/deviceshifu/deviceshifuhttp/deviceshifuhttp_test.go

```diff
@@ -186,3 +194,256 @@ func TestCreatehttpURIStringNoQuery(t *testing.T) {
+	assert.Nil(t, r.Error())
+
+	r = dc.Put().Do(context.TODO())
+	assert.Nil(t, r.Error())
+}
+
+func Test_commandHandleFuncHTTPCommandLine(t *testing.T) {
+	hs := mockHandlerServer()
+	defer hs.Close()
+	addr := strings.Split(hs.URL, "//")[1]
+
+	hc, err := mockRestClient(addr, "")
+	if err != nil {
+		t.Errorf("create handler client error: %s", err.Error())
+	}
+	mockHandlerHTTPCli := &DeviceCommandHandlerHTTPCommandline{
+		client: hc,
+		CommandlineHandlerMetadata: &CommandlineHandlerMetadata{
+			edgeDeviceSpec: v1alpha1.EdgeDeviceSpec{
+				Address: &addr,
+			},
+			instruction: "test_instruction",
+			properties:  mockDeviceShifuInstruction(),
+		},
+	}
+
+	ds := mockDeviceServer(mockHandlerHTTPCli)
+	defer ds.Close()
+	dc, err := mockRestClient(ds.URL, "testing")
+	if err != nil {
+		t.Errorf("create device client error: %s", err.Error())
+	}
+
+	// start device client testing
+	r := dc.Post().Param("timeout", "1").Param("stub_toleration", "1").Do(context.TODO())
+	assert.Nil(t, r.Error())
+
+	r = dc.Post().Param("timeout", "aa").Param("stub_toleration", "1").Do(context.TODO())
+	assert.Equal(t, "the server rejected our request for an unknown reason", r.Error().Error())
+
+	r = dc.Post().Param("timeout", "-1").Param("stub_toleration", "aa").Do(context.TODO())
+	assert.Equal(t, "the server rejected our request for an unknown reason", r.Error().Error())
+
+}
+
+func mockRestClient(url string, path string) (*rest.RESTClient, error) {
+	return rest.RESTClientFor(
+		&rest.Config{
+			Host:    url,
+			APIPath: path,
+			ContentConfig: rest.ContentConfig{
+				GroupVersion:         &v1.SchemeGroupVersion,
+				NegotiatedSerializer: scheme.Codecs.WithoutConversion(),
+			},
+			Username: "user",
+			Password: "pass",
+		},
+	)
+}
+
+type MockCommandHandler interface {
+	commandHandleFunc() http.HandlerFunc
+}
+
+func mockDeviceServer(h MockCommandHandler) *httptest.Server {
+	// catch device http request and response properly with specific paths
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		method := r.Method
+		if method == "GET" || method == "POST" {
+			path := r.URL.Path
+			switch path {
+			case "/testing/apps/v1":
+				println("ds get testing call, calling the handler server")
+				f := h.commandHandleFunc()
+				f.ServeHTTP(w, r)
+			default:
+				w.Header().Add("Content-Type", "application/json")
+				w.WriteHeader(http.StatusOK)
+				println("ds default request, path:", path)
+			}
+		} else {
+			println("invalid method")
+		}
+	}))
+	return server
+}
+
+func mockHandlerServer() *httptest.Server {
+	// catch handler http request and response properly with specific paths
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		switch path {
+		case "/test_instruction":
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			println("handler get the instruction and executed.")
```

Comment:
Maybe instead of print the result out, you can use assert here to make sure you received expected http requests, which would be more intuitive for unit tests.

Reply:
good idea